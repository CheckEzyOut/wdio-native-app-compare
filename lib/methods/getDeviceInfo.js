import { determineIphoneXSeries, determineLargeIphoneXSeries, getScreenshotSize } from '../helpers/utils'

let DEVICE_INFO = null

/**
 * Get the device info and store it to a global if needed
 *
 * @param   {string}  screenshot
 *
 * @return  {Promise<{
 *      dpr: number,
 *      screenSize: {
 *          height: number,
 *          width: number,
 *      },
 *      screenshotHeight: number,
 *      screenshotWidth: number,
 *      isIphoneXSeries: boolean,
 *      isLargeIphoneXSeries: boolean,
 * }>}
 */
export async function getDeviceInfo(screenshot) {
    if (DEVICE_INFO) {
        return DEVICE_INFO
    }
    const { height: screenshotHeight, width: screenshotWidth } = getScreenshotSize(screenshot)
    const screenSize = (await browser.getWindowSize())
    const isIphoneXSeries = determineIphoneXSeries(screenSize)
    const isLargeIphoneXSeries = determineLargeIphoneXSeries(screenSize)

    DEVICE_INFO = {
        dpr: screenshotWidth / screenSize.width,
        screenSize,
        screenshotHeight,
        screenshotWidth,
        isIphoneXSeries,
        isLargeIphoneXSeries,
    }

    return DEVICE_INFO
}
