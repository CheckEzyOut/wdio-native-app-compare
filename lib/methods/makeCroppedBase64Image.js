import { yellow } from 'chalk';
import { createWriteStream } from 'fs-extra';

const Canvas = require('canvas-prebuilt');

/**
 * Save a (cropped) canvas screen
 *
 * @TODO: also check right and bottom coordinates
 *
 * @param {buffer}  screenshot
 * @param {object}  rectangles
 * @param {object}  resizeDimensions
 */
export function makeCroppedBase64Image(screenshot, rectangles, resizeDimensions) {
    const CanvasImage = Canvas.Image;
    const image = new CanvasImage();
    let canvas;

    console.log('rectangles = ', rectangles);
    console.log('resizeDimensions = ', resizeDimensions);

    image.onerror = (err) => {
        throw err
    };

    image.onload = () => {
        const w = rectangles.width + resizeDimensions.left + resizeDimensions.right;
        const h = rectangles.height + resizeDimensions.top + resizeDimensions.bottom;
        canvas = new Canvas(w, h);
        const ctx = canvas.getContext('2d');

        let sourceXStart = rectangles.x - resizeDimensions.left;
        let sourceYStart = rectangles.y - resizeDimensions.top;

        if (sourceXStart < 0) {
            console.log(yellow(`
#####################################################################################
 THE RESIZE DIMENSION LEFT '${resizeDimensions.left}' MADE THE CROPPING GO OUT OF
 THE IMAGE BOUNDARIES RESULTING IN AN IMAGE STARTPOSITION '${sourceXStart}'.
 THIS HAS BEEN DEFAULTED TO '0'
#####################################################################################
`));
            sourceXStart = 0;
        }

        if (sourceYStart < 0) {
            console.log(yellow(`
#####################################################################################
 THE RESIZE DIMENSION LEFT '${resizeDimensions.top}' MADE THE CROPPING GO OUT OF
 THE IMAGE BOUNDARIES RESULTING IN AN IMAGE STARTPOSITION '${sourceYStart}'.
 THIS HAS BEEN DEFAULTED TO '0'
#####################################################################################
`));
            sourceYStart = 0;
        }

        ctx.drawImage(image,
            // Start at x/y pixels from the left and the top of the image (crop)
            sourceXStart, sourceYStart,
            // "Get" a (w * h) area from the source image (crop)
            w, h,
            // Place the result at 0, 0 in the canvas,
            0, 0,
            // With as width / height: 100 * 100 (scale)
            w, h
        );
    };
    image.src = `data:image/png;base64,${screenshot}`;

    return canvas.toDataURL().replace(/^data:image\/png;base64,/, '');
}
