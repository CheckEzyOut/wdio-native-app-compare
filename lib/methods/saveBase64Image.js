import { outputFile } from 'fs-extra';

/**
 * Save the base64 image to a file
 *
 * @param   {string}  base64Image
 * @param   {string}  filePath
 *
 * @returns {Promise<void>}
 */
export default async function saveBase64Image(base64Image, filePath) {
    return outputFile(filePath, base64Image, 'base64');
}
