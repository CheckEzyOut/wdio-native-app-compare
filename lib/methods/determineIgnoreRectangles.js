import {
    IPHONE_X_DEFAULT_HOME_BAR,
    IPHONE_X_LARGE_HOME_BAR,
    NAVIGATION_BAR_SELECTORS,
    STATUS_BAR_SELECTORS,
} from '../helpers/constants'
import { multiplyIosObjectValuesAgainstDPR } from '../helpers/utils'
import { getDeviceInfo } from './getDeviceInfo'
import { nativeAppCompareLog } from '../helpers/logger'

let NAVIGATION_BAR_DIMENSIONS = null
let STATUS_BAR_DIMENSIONS = null

/**
 * Determine the ignore rectangles
 *
 * @param {string}  base64Image
 * @param {object}  options
 *
 * @return {Promise<Array>}
 */
export async function determineIgnoreRectangles(base64Image, options) {
    let rectangles = []
    let allowInvisibleElements
    const deviceInfo = await getDeviceInfo(base64Image)

    // With Android and UiAutomator2 the status bar is not in the view by default
    // It can be seen when {settings: {allowInvisibleElements: true}} is set
    // See http://appium.io/docs/en/advanced-concepts/settings/
    if (browser.isAndroid && (options.blockOutStatusBar || options.blockOutNavigationBar || options.elementBlockOuts.length > 0)) {
        allowInvisibleElements = (await browser.getSettings()).allowInvisibleElements
        await browser.updateSettings({ allowInvisibleElements: true })
    }

    if (options.blockOutStatusBar) {
        rectangles.push(await determineStatusBarRectangles(deviceInfo.dpr, deviceInfo.screenshotWidth))
    }

    if (browser.isAndroid && options.blockOutNavigationBar) {
        rectangles.push(await determineNavigationBarRectangles(deviceInfo.dpr, deviceInfo.screenshotWidth))
    }

    if (options.blockOuts.length > 0) {
        rectangles = rectangles.concat(determineBlockOuts(options.blockOuts))
    }

    if (options.elementBlockOuts.length > 0) {
        rectangles = rectangles.concat(await determineElementBlockOuts(options.elementBlockOuts, deviceInfo))
    }

    if (deviceInfo.isIphoneXSeries && options.blockOutIphoneXBottomBar) {
        rectangles = rectangles.concat(deviceInfo.isLargeIphoneXSeries ? IPHONE_X_LARGE_HOME_BAR : IPHONE_X_DEFAULT_HOME_BAR)
    }

    if (browser.isAndroid && (options.blockOutStatusBar || options.blockOutNavigationBar || options.elementBlockOuts.length > 0)) {
        await browser.updateSettings({ allowInvisibleElements })
    }

    return rectangles
}

/**
 * Determine the status bar height for iOS/Android
 *
 * @param {number} dpr The device pixel ratio
 *
 * @return {Promise<{
 *      height: number,
 *      width: number,
 *      x: number,
 *      y: number
 * }>}
 */
async function determineStatusBarRectangles(dpr, width) {
    if (STATUS_BAR_DIMENSIONS) {
        return STATUS_BAR_DIMENSIONS
    }

    let rectangles

    try {
        rectangles = await getElementRectangles(await ($(STATUS_BAR_SELECTORS[ browser.isIOS ? 'IOS' : 'ANDROID' ])))
    } catch (e) {
        nativeAppCompareLog.warn('The statusbar could not be determined and has been set to the defaults!\n')

        //@TODO: default to an Android or iOS (iPhone (X)) height
        rectangles = { height: 26, width, x: 0, y: 0 }
    }

    STATUS_BAR_DIMENSIONS = multiplyIosObjectValuesAgainstDPR({
        height: rectangles.height,
        width: rectangles.width,
        x: 0,
        y: 0
    }, dpr)

    return STATUS_BAR_DIMENSIONS
}

/**
 * Determine the navigation bar height for iOS/Android
 *
 * @param {number} dpr The device pixel ratio
 *
 * @return {Promise<{
 *      height: number,
 *      width: number,
 *      x: number,
 *      y: number
 * }>}
 */
async function determineNavigationBarRectangles(dpr, width) {
    if (NAVIGATION_BAR_DIMENSIONS) {
        return NAVIGATION_BAR_DIMENSIONS
    }

    let rectangles

    try {
        rectangles = await getElementRectangles(await $(NAVIGATION_BAR_SELECTORS.ANDROID))
    } catch (e) {
        nativeAppCompareLog.warn('The navigation bar could not be determined and has been set to the defaults!\n')

        rectangles = { height: 0, width: 0, x: 0, y: 0 }
    }

    NAVIGATION_BAR_DIMENSIONS = multiplyIosObjectValuesAgainstDPR({
        height: rectangles.height,
        width: width,
        x: rectangles.x,
        y: rectangles.y
    }, dpr)

    return NAVIGATION_BAR_DIMENSIONS
}

/**
 * Determine the block outs data
 *
 * @param {array}   blockOuts
 *
 * @return {[{
 *      height: number,
 *      width: number,
 *      x: number,
 *      y: number
 * }]}
 */
function determineBlockOuts(blockOuts) {
    const dimensions = []
    const validKeys = [ 'x', 'y', 'width', 'height' ]

    blockOuts.forEach(blockOut => {
        let validBlockOutObject = true
        Object.keys(blockOut).forEach(key => {
            if (!validKeys.includes(key)) {
                nativeAppCompareLog.error(`
###################################################
 Blockout:
 ${ JSON.stringify(blockOut) }

 is not allowed. Key: ${ key } is not allowed, 
 only 'x', 'y', 'width', 'height' are allowed
###################################################
`)
                validBlockOutObject = false
            }
        })

        if (validBlockOutObject) {
            dimensions.push(blockOut)
        }
    })

    return dimensions
}

/**
 * Determine the element block out data
 *
 * @param {array}   components
 * @param {object}  deviceInfo the device info object
 *
 * @return {Promise<[{
 *      height: number,
 *      width: number,
 *      x: number,
 *      y: number
 * }]>}
 */
async function determineElementBlockOuts(components, deviceInfo) {
    const dimensions = []

    for (let component of components) {
        const selector = await component.element.selector
        // By providing elementNumber a specific element from the element array will be used
        const hasNumber = component.hasOwnProperty('elementNumber')

        try {
            const elements = hasNumber ? [ (await ($$(selector)))[ component.elementNumber ] ] : (await ($$(selector)))
            const margin = component.margin || 0
            for (let element of elements) {
                const rectangles = await getElementRectangles(element)
                dimensions.push(determineDimensions(rectangles, margin))
            }
        } catch (e) {
            nativeAppCompareLog.warn(`Something went wrong with finding / determining the rectangles of the element with selector: '${ selector }'.`)
            nativeAppCompareLog.warn('This error has been catched and shown below, but the test did not broke based on this error')
            nativeAppCompareLog.warn('\n', e, '\n')

        }
    }

    return dimensions.map(dimension => multiplyIosObjectValuesAgainstDPR(dimension, deviceInfo.dpr))
}

/**
 * Determine the dimensions
 *
 * @param {object} rectangles   An object with height, width,x and y
 * @param {number} margin       The margin that needs to be used
 *
 * @return {{
 *      height: number,
 *      width: number,
 *      x: number,
 *      y: number
 * }}
 */
function determineDimensions(rectangles, margin) {
    return {
        height: rectangles.height + (2 * margin),
        width: rectangles.width + (2 * margin),
        x: rectangles.x - margin,
        y: rectangles.y - margin,
    }
}

/**
 * Get the rectangles of an element
 *
 * @param {object}  element
 * @return {Promise<{
 *      height: number,
 *      width: number,
 *      x: number,
 *      y: number
 * }>}
 */
async function getElementRectangles(element) {
    return (await (browser.getElementRect(element.elementId)))
}
