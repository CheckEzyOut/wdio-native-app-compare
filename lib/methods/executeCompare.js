import { readFileSync } from 'fs-extra';
import { join, resolve } from 'path';
import { compareImages } from '../resemble/compareImages';
import saveBase64Image from './saveBase64Image';

/**
 * Compare 2 images with each other
 *
 * @param {object}  folders
 * @param {string}  fileName
 * @param {object}  options
 *
 * @return {Promise<{
 *      fileName: {string},
 *      folders: {object},
 *      misMatchPercentage: number
 * }>}
 */
export async function executeCompare(folders, fileName, options) {
    if (options.ignoreAlpha) {
        options.ignore.push('alpha')
    }
    if (options.ignoreAntialiasing) {
        options.ignore.push('antialiasing')
    }
    if (options.ignoreColors) {
        options.ignore.push('colors')
    }
    if (options.ignoreLess) {
        options.ignore.push('less')
    }
    if (options.ignoreNothing) {
        options.ignore.push('nothing')
    }

    // Execute the compare and retrieve the data
    const data = await compareImages(
        readFileSync(resolve(folders.baseline, fileName)),
        readFileSync(resolve(folders.actual, fileName)),
        options
    );
    const misMatchPercentage = options.rawMisMatchPercentage
        ? data.rawMisMatchPercentage
        : Number(data.rawMisMatchPercentage.toFixed(2));

    // Save the diff when there is a diff or when debug mode is on
    if (misMatchPercentage > options.saveAboveTolerance || options.debug) {
        await saveBase64Image(resolve(folders.diff, fileName), Buffer.from(data.getBuffer()).toString('base64'));
    }

    // Return the comparison data
    return {
        fileName,
        folders,
        misMatchPercentage,
    };
}
