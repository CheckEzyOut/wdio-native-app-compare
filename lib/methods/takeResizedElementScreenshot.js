import { DEFAULT_RESIZE_DIMENSIONS } from '../helpers/constants'
import { takeBase64Screenshot } from '../methods/takeBase64Screenshot'
import { getDeviceInfo } from './getDeviceInfo'
import { makeCroppedBase64Image } from './makeCroppedBase64Image'
import { getElementRectangles } from './rectangles'

/**
 * Save a resized element screenshot
 *
 * @param {object}  element
 * @param {string}  filePath
 * @param {object}  resizeDimensions
 *
 * @return {Promise<string>}
 *
 * @private
 */
export async function takeResizedElementScreenshot(element, filePath, resizeDimensions = DEFAULT_RESIZE_DIMENSIONS) {
    // Determine location and size of element
    const elementRectangles = await getElementRectangles(element)

    // Take a screenshot
    const base64Screenshot = await takeBase64Screenshot()

    // Get the device info like screenSize, screenshotHeight, screenshotHeight
    const deviceInfo = await getDeviceInfo(base64Screenshot)

    // Create the cropped screenshot and return it
    return await makeCroppedBase64Image(
        base64Screenshot,
        {
            height: (elementRectangles.height * deviceInfo.dpr),
            width: (elementRectangles.width * deviceInfo.dpr),
            x: (elementRectangles.x * deviceInfo.dpr),
            y: (elementRectangles.y * deviceInfo.dpr),
        },
        resizeDimensions,
    )
}
