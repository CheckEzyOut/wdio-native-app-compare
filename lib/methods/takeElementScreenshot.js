import { yellow } from 'chalk';
import { takeResizedElementScreenshot } from './takeResizedElementScreenshot';


/**
 * Take the element screenshot based on the elementIdScreenshot of webdriver.io
 *
 * @param {object}  element
 * @param {string}  filePath
 *
 * @return {Promise<void>}
 *
 * @private
 */
export async function takeElementScreenshot(element, filePath) {
    const id = await element.value.ELEMENT;

    try {
        return (await browser.elementIdScreenshot(id)).value;
    } catch (e) {
        console.log(yellow(`
#####################################################################################
 AN ELEMENT SCREENSHOT COULD NOT BE TAKEN WITH THE NEW 'elementScreenshot()' METHOD, 
 A FALLBACK HAS BEEN EXECUTED AND THE ELEMENT WILL BE SAVED.
 USE 'automationName: "UiAutomator2"' FOR ANDROID TO FULLY USE THE POWER OF APPIUM
#####################################################################################
`));
        return takeResizedElementScreenshot(element, filePath);
    }
}
