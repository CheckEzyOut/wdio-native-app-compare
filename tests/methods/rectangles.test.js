import * as DeviceInfo from '../../lib/methods/getDeviceInfo'
import {IMAGE_STRING} from '../mocks/mocks'

describe('rectangles', () => {
    let getDeviceInfoSpy, ignoreRectanglesOptions, Rectangles

    beforeEach(() => {
        jest.isolateModules(() => {
            Rectangles = require('../../lib/methods/rectangles')
        })
        delete global.browser
        delete global.$$

        global.$$ = () => [{'element-selector-1': {}}]
        global.browser = {
            getElementRect: jest.fn().mockResolvedValue({x: 1, y: 2, width: 10, height: 20}),
            getSettings: jest.fn().mockResolvedValue({allowInvisibleElements: false}),
            updateSettings: jest.fn().mockResolvedValue({}),
            isIOS: false,
            isAndroid: false,
        }
        getDeviceInfoSpy = jest.spyOn(DeviceInfo, 'getDeviceInfo').mockResolvedValue({
            dpr: 2,
            isIphoneXSeries: true,
            isLargeIphoneXSeries: false,
            screenSize: {
                height: 1700,
                width: 812,
            },
            screenshotHeight: 1700,
            screenshotWidth: 812,
            isPortrait: true,
            iosRectangles: {
                STATUS_BAR: 20,
                HOME_BAR: {
                    PORTRAIT: {
                        height: 0,
                        width: 0,
                        x: 0,
                        y: 0,
                    },
                },
            },
        })
        ignoreRectanglesOptions = {
            blockOuts: [],
            blockOutNavigationBar: false,
            blockOutStatusBar: false,
            elementBlockOuts: [],
        }
    })

    afterEach(() => {
        global.browser = {
            getElementRect: jest.fn().mockRestore(),
            getSettings: jest.fn().mockRestore(),
            updateSettings: jest.fn().mockRestore(),
        }
        getDeviceInfoSpy.mockRestore()
    })

    describe('determineIgnoreRectangles', () => {
        it('should return nothing if no ignore rectangles need to be returned', async () => {
            expect(await Rectangles.determineIgnoreRectangles(IMAGE_STRING, ignoreRectanglesOptions)).toMatchSnapshot()
        })

        it('should get the status bar block outs', async () => {
            ignoreRectanglesOptions.blockOutStatusBar = true

            expect(await Rectangles.determineIgnoreRectangles(IMAGE_STRING, ignoreRectanglesOptions)).toMatchSnapshot()
        })

        it('should get the android navigation bar block outs', async () => {
            global.browser.isAndroid = true
            ignoreRectanglesOptions.blockOutNavigationBar = true

            expect(await Rectangles.determineIgnoreRectangles(IMAGE_STRING, ignoreRectanglesOptions)).toMatchSnapshot()
        })

        it('should get the block outs', async () => {
            ignoreRectanglesOptions.blockOuts = [{x: 11, y: 22, width: 33, height: 44}, {
                x: 21,
                y: 32,
                width: 43,
                height: 54
            }]

            expect(await Rectangles.determineIgnoreRectangles(IMAGE_STRING, ignoreRectanglesOptions)).toMatchSnapshot()
        })

        it('should return iPhone X bottom bar rectangles', async () => {
            ignoreRectanglesOptions.blockOutIphoneXBottomBar = true
            getDeviceInfoSpy = jest.spyOn(DeviceInfo, 'getDeviceInfo').mockResolvedValue({
                dpr: 2,
                screenshotWidth: 812,
                isIphoneXSeries: true,
                isLargeIphoneXSeries: false,
                iosRectangles:{
                    STATUS_BAR: 44,
                    HOME_BAR: {
                        PORTRAIT: {
                            height: 5,
                            width: 148,
                            x: 133,
                            y: 883,
                        },
                    },
                }
            })
            global.browser.isIOS = true

            expect(await Rectangles.determineIgnoreRectangles(IMAGE_STRING, ignoreRectanglesOptions)).toMatchSnapshot()
        })

        it('should call the settings API for android when the status bar needs to be blocked out', async () => {
            global.browser.isAndroid = true
            ignoreRectanglesOptions.blockOutStatusBar = true

            await Rectangles.determineIgnoreRectangles(IMAGE_STRING, ignoreRectanglesOptions)

            expect(global.browser.getSettings).toBeCalled()
            expect(global.browser.updateSettings).toBeCalledWith({allowInvisibleElements: true})
            expect(global.browser.updateSettings).toHaveBeenLastCalledWith({allowInvisibleElements: false})
        })

        it('should call the settings API for android when the navigation bar needs to be blocked out', async () => {
            global.browser.isAndroid = true
            ignoreRectanglesOptions.blockOutNavigationBar = true

            await Rectangles.determineIgnoreRectangles(IMAGE_STRING, ignoreRectanglesOptions)

            expect(global.browser.getSettings).toBeCalled()
            expect(global.browser.updateSettings).toBeCalledWith({allowInvisibleElements: true})
            expect(global.browser.updateSettings).toHaveBeenLastCalledWith({allowInvisibleElements: false})
        })

        it('should call the settings API for android when elements needs to be blocked out', async () => {
            global.browser.isAndroid = true
            ignoreRectanglesOptions.elementBlockOuts = [{element: {selector: 'selector-1'}}, {
                element: {selector: 'selector-1'}
            }]
            await Rectangles.determineIgnoreRectangles(IMAGE_STRING, ignoreRectanglesOptions)
            expect(global.browser.getSettings).toBeCalled()
            expect(global.browser.updateSettings).toBeCalledWith({allowInvisibleElements: true})
            expect(global.browser.updateSettings).toHaveBeenLastCalledWith({allowInvisibleElements: false})
        })
    })

    describe('determineStatusBarRectangles', () => {
        it('should determine the initial statusbar rectangles', async () => {
            expect(await Rectangles.determineStatusBarRectangles({
                dpr: 4,
                iosStatusBar: 20,
                screenshotWidth: 300,
            })).toMatchSnapshot()
        })

        it('should determine the statusbar rectangles by using the already defined one', async () => {
            expect(await Rectangles.determineStatusBarRectangles({
                dpr: 10,
                iosStatusBar: 20,
                screenshotWidth: 30,
            })).toMatchSnapshot()

            // Call it again, this should not alter the outcome
            expect(await Rectangles.determineStatusBarRectangles({
                dpr: 40,
                iosStatusBar: 50,
                screenshotWidth: 60,
            })).toMatchSnapshot()
        })

        it('should determine the iOS statusbar rectangles', async () => {
            global.$ = () => [{'element-selector-1': {}}]
            global.browser.isIOS = true

            expect(await Rectangles.determineStatusBarRectangles({
                dpr: 4,
                iosStatusBar: 20,
                screenshotWidth: 300,
            })).toMatchSnapshot()
        })
    })

    describe('determineAndroidNavigationBarRectangles', () => {
        it('should determine the navigationbar rectangles', async () => {
            global.$ = () => [{'element-selector-1': {}}]

            expect(await Rectangles.determineAndroidNavigationBarRectangles({screenshotWidth: 500})).toMatchSnapshot()
        })

        it('should determine the navigationbar rectangles by using the already defined one', async () => {
            global.$ = () => [{'element-selector-1': {}}]

            expect(await Rectangles.determineAndroidNavigationBarRectangles({screenshotWidth: 500})).toMatchSnapshot()
            // Call it again, this should not alter the outcome
            expect(await Rectangles.determineAndroidNavigationBarRectangles({screenshotWidth: 1000})).toMatchSnapshot()
        })
    })

    describe('determineValidBlockOuts', () => {
        it('should return the valid blockouts if an valid array if provided', () => {
            const blockOuts = [
                {x: 1, y: 2, width: 3, height: 4},
                {x: 1, y: 2, width: 3, height: 4},
            ]

            expect(Rectangles.determineValidBlockOuts(blockOuts)).toMatchSnapshot()
        })
        it('should return the valid blockouts if an array with invalid keys is provided', () => {
            const blockOuts = [
                {x: 5, y: 6, width: 7, height: 8},
                {b: 5, y: 6, widthy: 7, height: 8},
                {x: 9, y: 12, width: 13, height: 14},
                {x: 9, ho: 12, width: 13, bar: 14},
            ]

            expect(Rectangles.determineValidBlockOuts(blockOuts)).toMatchSnapshot()
        })
    })

    describe('determineElementBlockOuts', () => {
        it('should return an empty array if no elements are provided', async () => {
            expect(await Rectangles.determineElementBlockOuts({
                components: [],
                dpr: 1,
            })).toMatchSnapshot()
        })

        it('should return an array of one element blockout if 1 element is provided', async () => {
            const selector = 'element-selector-1'
            const components = [
                {element: {selector}}
            ]
            global.$$ = () => [{'element-selector-1': {}}]

            expect(await Rectangles.determineElementBlockOuts({
                components: components,
                dpr: 1,
            })).toMatchSnapshot()
        })

        it('should return an array of one element blockout if multiple elements are provided but 1 with an element number', async () => {
            const selector = 'element-selector-1'
            const components = [
                {element: {selector}, elementNumber: 1}
            ]
            global.$$ = () => [{'element-selector-1': {}}, {'element-selector-1': {}}, {'element-selector-1': {}}]

            expect(await Rectangles.determineElementBlockOuts({
                components: components,
                dpr: 1,
            })).toMatchSnapshot()
        })

        it('should return an array of element blockouts if multiple elements are provided', async () => {
            const selector = 'element-selector-1'
            const components = [{element: {selector}}, {element: {selector}}, {element: {selector}}]
            global.$$ = () => [{'element-selector-1': {}}, {'element-selector-1': {}}, {'element-selector-1': {}}]

            expect(await Rectangles.determineElementBlockOuts({
                components: components,
                dpr: 1,
            })).toMatchSnapshot()
        })
    })

    describe('determineDimensions', () => {
        it('should be able to determine the dimensions when no margin is provided', () => {
            expect(Rectangles.determineDimensions({
                x: 100,
                y: 200,
                width: 1000,
                height: 2000,
            }, 0)).toMatchSnapshot()
        })

        it('should be able to determine the dimensions when a positive margin is provided', () => {
            expect(Rectangles.determineDimensions({
                x: 100,
                y: 200,
                width: 1000,
                height: 2000,
            }, 25)).toMatchSnapshot()
        })

        it('should be able to determine the dimensions when a negative margin is provided', () => {
            expect(Rectangles.determineDimensions({
                x: 100,
                y: 200,
                width: 1000,
                height: 2000,
            }, -25)).toMatchSnapshot()
        })
    })

    describe('getElementRectangles', () => {
        it('should return element rectangles', async () => {

            expect(await Rectangles.getElementRectangles({elementId: 1})).toMatchSnapshot()
            expect(global.browser.getElementRect).toBeCalledWith(1)
        })
    })
})
