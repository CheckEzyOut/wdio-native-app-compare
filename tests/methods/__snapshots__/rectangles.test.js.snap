// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rectangles determineDimensions should be able to determine the dimensions when a negative margin is provided 1`] = `
Object {
  "height": 1950,
  "width": 950,
  "x": 125,
  "y": 225,
}
`;

exports[`rectangles determineDimensions should be able to determine the dimensions when a positive margin is provided 1`] = `
Object {
  "height": 2050,
  "width": 1050,
  "x": 75,
  "y": 175,
}
`;

exports[`rectangles determineDimensions should be able to determine the dimensions when no margin is provided 1`] = `
Object {
  "height": 2000,
  "width": 1000,
  "x": 100,
  "y": 200,
}
`;

exports[`rectangles determineElementBlockOuts should return an array of element blockouts if multiple elements are provided 1`] = `
Array [
  Object {
    "height": 20,
    "width": 10,
    "x": 1,
    "y": 2,
  },
  Object {
    "height": 20,
    "width": 10,
    "x": 1,
    "y": 2,
  },
  Object {
    "height": 20,
    "width": 10,
    "x": 1,
    "y": 2,
  },
  Object {
    "height": 20,
    "width": 10,
    "x": 1,
    "y": 2,
  },
  Object {
    "height": 20,
    "width": 10,
    "x": 1,
    "y": 2,
  },
  Object {
    "height": 20,
    "width": 10,
    "x": 1,
    "y": 2,
  },
  Object {
    "height": 20,
    "width": 10,
    "x": 1,
    "y": 2,
  },
  Object {
    "height": 20,
    "width": 10,
    "x": 1,
    "y": 2,
  },
  Object {
    "height": 20,
    "width": 10,
    "x": 1,
    "y": 2,
  },
]
`;

exports[`rectangles determineElementBlockOuts should return an array of one element blockout if 1 element is provided 1`] = `
Array [
  Object {
    "height": 20,
    "width": 10,
    "x": 1,
    "y": 2,
  },
]
`;

exports[`rectangles determineElementBlockOuts should return an array of one element blockout if multiple elements are provided but 1 with an element number 1`] = `
Array [
  Object {
    "height": 20,
    "width": 10,
    "x": 1,
    "y": 2,
  },
]
`;

exports[`rectangles determineElementBlockOuts should return an empty array if no elements are provided 1`] = `Array []`;

exports[`rectangles determineIgnoreRectangles should get the android navigation bar block outs 1`] = `
Array [
  Object {
    "height": 0,
    "width": 812,
    "x": 0,
    "y": 0,
  },
]
`;

exports[`rectangles determineIgnoreRectangles should get the block outs 1`] = `
Array [
  Object {
    "height": 44,
    "width": 33,
    "x": 11,
    "y": 22,
  },
  Object {
    "height": 54,
    "width": 43,
    "x": 21,
    "y": 32,
  },
]
`;

exports[`rectangles determineIgnoreRectangles should get the status bar block outs 1`] = `
Array [
  Object {
    "height": 26,
    "width": 812,
    "x": 0,
    "y": 0,
  },
]
`;

exports[`rectangles determineIgnoreRectangles should return iPhone X Max bottom bar rectangles 1`] = `
Array [
  Object {
    "height": 15,
    "width": 444,
    "x": 399,
    "y": 2649,
  },
]
`;

exports[`rectangles determineIgnoreRectangles should return iPhone X regular bottom bar rectangles 1`] = `
Array [
  Object {
    "height": 15,
    "width": 402,
    "x": 362,
    "y": 2397,
  },
]
`;

exports[`rectangles determineIgnoreRectangles should return nothing if no ignore rectangles need to be returned 1`] = `Array []`;

exports[`rectangles determineNavigationBarRectangles should determine the navigationbar rectangles 1`] = `
Object {
  "height": 0,
  "width": 812,
  "x": 0,
  "y": 0,
}
`;

exports[`rectangles determineNavigationBarRectangles should determine the navigationbar rectangles 2`] = `
Object {
  "height": 0,
  "width": 812,
  "x": 0,
  "y": 0,
}
`;

exports[`rectangles determineStatusBarRectangles should determine the statusbar rectangles 1`] = `
Object {
  "height": 26,
  "width": 812,
  "x": 0,
  "y": 0,
}
`;

exports[`rectangles determineStatusBarRectangles should determine the statusbar rectangles 2`] = `
Object {
  "height": 26,
  "width": 812,
  "x": 0,
  "y": 0,
}
`;

exports[`rectangles determineValidBlockOuts should return the valid blockouts if an array with invalid keys is provided 1`] = `
Array [
  Object {
    "height": 8,
    "width": 7,
    "x": 5,
    "y": 6,
  },
  Object {
    "height": 14,
    "width": 13,
    "x": 9,
    "y": 12,
  },
]
`;

exports[`rectangles determineValidBlockOuts should return the valid blockouts if an valid array if provided 1`] = `
Array [
  Object {
    "height": 4,
    "width": 3,
    "x": 1,
    "y": 2,
  },
  Object {
    "height": 4,
    "width": 3,
    "x": 1,
    "y": 2,
  },
]
`;

exports[`rectangles getElementRectangles should return element rectangles 1`] = `
Object {
  "height": 20,
  "width": 10,
  "x": 1,
  "y": 2,
}
`;
